### ISSUE TITLE
End-to-end testing with PHP backend and JS Frontend

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#1335]
## The problem

The code coverage of the JS tests is lacking…

> Unit tests: This could have been detected by JS unit tests, but those are much weaker than our PHP ones and only cover about [62%](https://privatebin.info/jscoverage/) of the logic, compared to [85%](https://privatebin.info/coverage/) (most modules are above 95%, but the cloud storage backends are badly or not at all tested, dragging the score down). In particular, [the problematic line is only covered one time](https://privatebin.info/jscoverage/privatebin.js.html#L4374) and the [bootstrap(3) specific event handler isn't covered at all](https://privatebin.info/jscoverage/privatebin.js.html#L3697).

## The solution

* it must work offline (possibly in a container or whatever)
* and be fully automated
* must run automatically in GitHub Actions
* for a start should cover the most common scenarios [which?], later more could be added

Ideas would be:
* cypress?

## Alternatives

* manual testing before a release like currently [suggested in the release checklist](https://github.com/PrivateBin/PrivateBin/wiki/Release-Checklist)

## Additional context

This issue has been discovered [as a lessons learned from previous releases](https://github.com/orgs/PrivateBin/discussions/1334#discussioncomment-9428293), which often had bugs being introduced. To have more confidence on code changes etc.

**The important bit:** Dear reader, if you feel like this is a thing you could contribute knowledge or actually doing it, feel free to comment here and help! This will greatly help the PrivateBin project.

**Comment by Ribas160** [comment]
Adding e2e test to the project is a great idea I think!
I did some research on this topic and found a [thread on Reddit](https://www.reddit.com/r/webdev/comments/1aryiie/best_e2e_testing_framework/). In this thread a [Playwright](https://playwright.dev/) framework is highly recommended, also compared to [Cypress](https://www.cypress.io/). I checked the github repos of these frameworks [microsoft/playwright](https://github.com/microsoft/playwright), [cypress-io/cypress](https://github.com/cypress-io/cypress) and Playwright has much more stars, also in npm packages statistic [package/playwright](https://www.npmjs.com/package/playwright), [package/cypress](https://www.npmjs.com/package/cypress) we could see that Playwright has much more downloads and it's popularity is glowing, but Cypress is kind of stagnant right now.
Also the [Playwright syntax](https://playwright.dev/docs/writing-tests#first-test) looks simplier for me copared to [Cypress syntax](https://docs.cypress.io/app/end-to-end-testing/writing-your-first-end-to-end-test#Step-3-Click-an-element).
So I think Playwright might be a better choice for e2e testing framework, but anyway if anyone has another opinion feel free to share it.

Also I think we should add a command to the [privatebin/unit-testing](https://github.com/PrivateBin/docker-unit-testing) docker image, it will help us to automate these kind of tests.

**Comment by Ribas160** [comment]
@rugk 
Let's think how these tests may look like

I think that for the beginning we can start from some simple tests, for example:
1. Enter some random data to the `textarea` and create the new paste.
2. Enter some random data to the `textarea` and press the "Preview" button.
3. Create new paste with random data and press "New" button.
4. Create new paste with random data and press "Clone" button.
5. Create new paste with enabled discussions and send a few comments.

What do you think about it? Is my point of view on e2e tests scenarios correct in your opinion?

**Comment by rugk** [comment]
Yeah, I think the actual use cases to test are pretty simple/or even trivial and your list is a good start. Later we'd likely want to add _receiving_ a paste (rather than just generating it)… or actually a full run: generate -> receive -> delete (potentially) -> try to receive again (fail); maybe even with different options like "expire after read" etc.

But that all said, starting small is a very good idea! The real challenge seems to be finding the correct technology, so if you want to give some e2e testing framework a try… feel free!

**Comment by rugk** [comment]
From having a quick look, BTW, playwright indeed looks cool [e.g. with a very simple GitHub Actions integration](https://playwright.dev/docs/ci-intro).

**Comment by Ribas160** [comment]
@rugk
For testing I wrote some e2e tests and ran into some issues.

First issue is that tests fail because of 10 second timeout between paste creation. I had to disable it to get all tests to pass.
https://github.com/PrivateBin/PrivateBin/blob/d220ba7c2ba69b138cfa82d8d3129a0d1bc2af23/lib/Configuration.php#L91-L96

Second issue is that the project needs to be running during tests, and first command is [go to project url](https://playwright.dev/docs/writing-tests#first-test)

On my local machine I just run project on localhost and write
```
await page.goto('http://localhost');
```

But what have I wright for running tests on the PrivateBin's CI?

**Comment by elrido** [comment]
For the traffic limiter, the same issue occurs in the php unit tests and there we use the unit test `Helper` class to set the limit to 0 for tests that need to post repeatedly:
https://github.com/PrivateBin/PrivateBin/blob/d220ba7c2ba69b138cfa82d8d3129a0d1bc2af23/tst/ControllerTest.php#L170-L172

ATM the CI on github doesn't spin up any server, so you'd have to install and configure a webserver and php process manager in there. Presumably you'd also needed to install a browser to drive the end-to-end tests from? You could (ab)use one of the docker images bind-mounting the locally checked out git repo into it's `/srv/www`?

TBH, this complexity is why I avoided these types of tests and stuck to writing simpler (to me) unit tests, where I'd test either server or client, based on mock requests or mock responses.

**Comment by Ribas160** [comment]
> For the traffic limiter, the same issue occurs in the php unit tests and there we use the unit test Helper class to set the limit to 0 for tests that need to post repeatedly

The problem is tests are running in node environment, so it can not manipulate by php unit test `Helper`.

> ATM the CI on github doesn't spin up any server, so you'd have to install and configure a webserver and php process manager in there.

As I see, we have to run a docker container with PrivateBin during the CI

> Presumably you'd also needed to install a browser to drive the end-to-end tests from?

Playwright [installs](https://playwright.dev/docs/browsers) all browsers it needs by default, so generally we have to just run the `npx playwright install` command.


> TBH, this complexity is why I avoided these types of tests and stuck to writing simpler (to me) unit tests, where I'd test either server or client, based on mock requests or mock responses.

You right that running these kind of tests is more complex than usual unit tests. Are they worth it? Which profit will we get apart of the experience?
